### PR TITLE
[ADD] udes_stock: Add get_movelines_from_quant function

### PR DESCRIPTION
**Pull request by peterpratt13**
User-story: 11545

Signed-off-by: Pete Pratt <peter.pratt@unipart.io>

### CODE REVIEW

File: addons/udes_stock/models/stock_quant.py [hunk 1]
```diff
@@ -136,3 +137,59 @@ def get_info(self):
     def quant_location_policy(self):
         self.ensure_one()
         self.location_id.apply_quant_policy()
+
+    def get_movelines_from_quant(self, aux_domain=None):
```
Comment (squah-u) [inline addons/udes_stock/models/stock_quant.py:141]: nit: `move_lines`

File: addons/udes_stock/models/stock_quant.py [hunk 2]
```diff
@@ -136,3 +137,59 @@ def get_info(self):
...
+        """Get the movelines associated to a quant
+        :param aux_domain: Extra domain arguments to add to the search
+        :returns: Associated movelines
```
Comment (squah-u) [inline addons/udes_stock/models/stock_quant.py:144]: nit: `move lines`

File: addons/udes_stock/models/stock_quant.py [hunk 3]
```diff
@@ -136,3 +137,59 @@ def get_info(self):
     def quant_location_policy(self):
         self.ensure_one()
         self.location_id.apply_quant_policy()
+
+    def get_movelines_from_quant(self, aux_domain=None):
+        """Get the movelines associated to a quant
```
Comment (squah-u) [inline addons/udes_stock/models/stock_quant.py:142]: nit: `move lines`

Comment (squah-u) [reply addons/udes_stock/models/stock_quant.py:142]: typo: `associated with`

File: addons/udes_stock/models/stock_quant.py [hunk 4]
```diff
@@ -136,3 +137,59 @@ def get_info(self):
...
+        if aux_domain is not None:
+            domain.extend(aux_domain)
+        return MoveLine.search(domain)
+
+    @api.multi
+    def _split(self, qty):
+        """Split a quant into two quants, a new quant with quantity `qty` and the original quantity of
+        self reduced by qty. If qty > quantity, it just returns self (original quant).
+        It gives the new quantity priority over reserved quantities
+        :param qty: Quantity to split from self, cretaing a new quant of quantity `qty`
```
Comment (squah-u) [inline addons/udes_stock/models/stock_quant.py:164]: typo: `creating`

File: addons/udes_stock/tests/test_stock_quant.py [hunk 5]
```diff
@@ -23,6 +23,9 @@ def setUpClass(cls):
         cls.quant2 = cls.create_quant(cls.apple.id, cls.test_location_01.id, 1.0)
         cls.quant3 = cls.create_quant(cls.apple.id, cls.test_location_01.id, 1.0)
 
+        # Create a banana quant for testing split
+        cls.quant3 = cls.create_quant(cls.banana.id, cls.test_location_01.id, 5.0)
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:27]: `quant3` already exists 3 lines up

File: addons/udes_stock/models/stock_quant.py [hunk 6]
```diff
@@ -136,3 +137,59 @@ def get_info(self):
...
+                if new_qty_round_reserved > 0
+                else reserved_quantity,
+            }
+        )
+        self.sudo().write(
+            {
+                "quantity": qty_round,
+                "reserved_quantity": qty_round if new_qty_round_reserved > 0 else 0,
+            }
+        )
```
Comment (squah-u) [inline addons/udes_stock/models/stock_quant.py:194]: The handling of `reserved_quantity` does not look right.  

If 1 item is split from a quant of 5, with 3 reserved, we end up with:
```
{ qty: 4, reserved: 4 }
{ qty: 1, reserved: 1 }
```

File: addons/udes_stock/tests/test_stock_quant.py [hunk 7]
```diff
@@ -96,3 +99,114 @@ def test_fifo_with_nones(self):
...
+        """Test get_mls_from_quant basic functionality"""
+        picking = self.create_picking(
+            self.picking_type_pick,
+            products_info=[{"product": self.apple, "qty": 2}],
+            confirm=True,
+            assign=True,
+        )
+        picking_mls = picking.mapped("move_line_ids")
+        mls_quant1 = self.quant1.get_movelines_from_quant()
+        mls_quant2 = self.quant1.get_movelines_from_quant()
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:113]: typo: `self.quant2`

File: addons/udes_stock/tests/test_stock_quant.py [hunk 8]
```diff
@@ -96,3 +99,114 @@ def test_fifo_with_nones(self):
...
+        self.assertEqual(self.quant.quantity, 5)
+        self.assertEqual(self.quant.reserved_quantity, 3)
+
+        # Split the quantity
+        new_quant = self.quant._split(2)
+        self.assertEqual(new_quant.quantity, 2)
+        self.assertEqual(new_quant.reserved_quantity, 2)
+        self.assertEqual(self.quant.quantity, 3)
+        self.assertEqual(self.quant.reserved_quantity, 1)
+        self.assertEqual(self.quant.quantity, 3)
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:182]: duplicate assert

File: addons/udes_stock/tests/test_stock_quant.py [hunk 9]
```diff
@@ -96,3 +99,114 @@ def test_fifo_with_nones(self):
...
+        new_quant = self.quant._split(4)
+        self.assertEqual(new_quant.quantity, 4)
+        self.assertEqual(new_quant.reserved_quantity, 3)
+        self.assertEqual(self.quant.quantity, 1)
+        self.assertEqual(self.quant.reserved_quantity, 0)
+
+        # Check the movelines have the quants
+        mls = picking.mapped("move_line_ids")
+        # TODO: Kevin's story 11558 should fix this
+        # self.assertEqual(mls.get_quants(), self.quant)
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:211]: 11558 shouldn't affect this?

File: addons/udes_stock/tests/test_stock_quant.py [hunk 10]
```diff
@@ -96,3 +99,114 @@ def test_fifo_with_nones(self):
...
+        self.assertEqual(self.quant.reserved_quantity, 5)
+
+        # Split the quantity
+        new_quant = self.quant._split(2)
+        self.assertEqual(new_quant.reserved_quantity, 3)
+        self.assertEqual(new_quant.quantity, 3)
+        self.assertEqual(self.quant.reserved_quantity, 2)
+        self.assertEqual(self.quant.quantity, 2)
+
+        # Check the movelines have the quants
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:159]: nit: `move lines`

File: addons/udes_stock/tests/test_stock_quant.py [hunk 11]
```diff
@@ -96,3 +99,114 @@ def test_fifo_with_nones(self):
...
+
+        # Split the quantity
+        new_quant = self.quant._split(2)
+        self.assertEqual(new_quant.quantity, 2)
+        self.assertEqual(new_quant.reserved_quantity, 2)
+        self.assertEqual(self.quant.quantity, 3)
+        self.assertEqual(self.quant.reserved_quantity, 1)
+        self.assertEqual(self.quant.quantity, 3)
+
+        # Check the movelines have the quants
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:184]: nit: `move lines`

File: addons/udes_stock/tests/test_stock_quant.py [hunk 12]
```diff
@@ -96,3 +99,114 @@ def test_fifo_with_nones(self):
...
+        self.assertEqual(self.quant.reserved_quantity, 3)
+
+        # Split the quantity
+        new_quant = self.quant._split(4)
+        self.assertEqual(new_quant.quantity, 4)
+        self.assertEqual(new_quant.reserved_quantity, 3)
+        self.assertEqual(self.quant.quantity, 1)
+        self.assertEqual(self.quant.reserved_quantity, 0)
+
+        # Check the movelines have the quants
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:208]: nit: `move lines`

File: addons/udes_stock/tests/test_stock_quant.py [hunk 13]
```diff
@@ -96,3 +99,114 @@ def test_fifo_with_nones(self):
...
+        mls = picking.mapped("move_line_ids")
+        # TODO: Kevin's story 11558 should fix this
+        # self.assertEqual(mls.get_quants(), self.quant)
+        self.assertEqual(mls.get_quants().filtered(lambda q: q.reserved_quantity > 0), self.quant)
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:212]: This looks redundant when there's already an assert for `self.quant`'s `reserved_quantity` above

File: addons/udes_stock/tests/test_stock_quant.py (hunk 9 above)
Comment (peterpratt13) [reply addons/udes_stock/tests/test_stock_quant.py]: No not here, in the other part of the story it `should` mean that quants get done in the correct order on action done. 
I've deleted teh mls.get_quants checks as it doesn't offer anything here

File: addons/udes_stock/tests/test_stock_quant.py [hunk 14]
```diff
@@ -96,3 +99,92 @@ def test_fifo_with_nones(self):
...
+        )
+        self.assertEqual(self.quant4.quantity, 5)
+        self.assertEqual(self.quant4.reserved_quantity, 3)
+
+        # Split the quantity
+        new_quant = self.quant4._split(4)
+        self.assertEqual(new_quant.quantity, 4)
+        self.assertEqual(new_quant.reserved_quantity, 3)
+        self.assertEqual(self.quant4.quantity, 1)
+        self.assertEqual(self.quant4.reserved_quantity, 0)
```
Comment (squah-u) [inline addons/udes_stock/tests/test_stock_quant.py:190]: nit: missing line break at end of file

**Comment by squah-u** [comment]
There's 1 more typo remaining inside the "3 hidden conversations" section

also remember to update the commit message to say `get_move_lines_from_quant` when squashing

**Comment by peterpratt13** [comment]
Thanks :) Typos everywhere, I've changed the function name  to just get_move_lines, to be consistent with get_quants in stock_move_line.py, you opposed to this?

**Comment by notcake** [comment]
Yeah that's good, as long as the commit message matches the code

**Comment by squah-u** [comment]
Yeah I meant to say `get_move_lines`.

**Review by squah-u — APPROVED** [review]
LGTM!

File: addons/udes_stock/tests/test_stock_quant.py [hunk 15]
```diff
@@ -23,6 +23,12 @@ def setUpClass(cls):
         cls.quant2 = cls.create_quant(cls.apple.id, cls.test_location_01.id, 1.0)
         cls.quant3 = cls.create_quant(cls.apple.id, cls.test_location_01.id, 1.0)
 
+<<<<<<< HEAD
```
Comment (ws-mps) [inline addons/udes_stock/tests/test_stock_quant.py]: This commit contains unresolved conflicts.

File: addons/udes_stock/tests/test_stock_quant.py [hunk 16]
```diff
@@ -96,3 +102,117 @@ def test_fifo_with_nones(self):
...
+            products_info=[{"product": self.banana, "qty": 1}],
+            confirm=True,
+            assign=True,
+        )
+        picking1_mls = picking1.mapped("move_line_ids")
+        picking2_mls = picking2.mapped("move_line_ids")
+        mls1 = self.quant4.get_move_lines(aux_domain=[("picking_id", "=", picking1.id)])
+        mls2 = self.quant4.get_move_lines(aux_domain=[("picking_id", "=", picking2.id)])
+        self.assertEqual(mls1, picking1_mls)
+<<<<<<< HEAD
```
Comment (ws-mps) [inline addons/udes_stock/tests/test_stock_quant.py]: Another unresolved conflict.

File: addons/udes_stock/tests/test_stock_quant.py [hunk 17]
```diff
@@ -96,3 +99,92 @@ def test_fifo_with_nones(self):
...
+
+    def test_get_mls_from_quant_basic(self):
+        """Test get_mls_from_quant basic functionality"""
+        picking = self.create_picking(
+            self.picking_type_pick,
+            products_info=[{"product": self.apple, "qty": 2}],
+            confirm=True,
+            assign=True,
+        )
+        picking_mls = picking.mapped("move_line_ids")
```
Comment (ws-mps) [inline addons/udes_stock/tests/test_stock_quant.py:111]: Do we have one or two mls here?

**Review by unirho — APPROVED** [review]

**Pull request merged**